File: Stacks and Queues/stackqueue.java
import java.util.*;

class stackqueue{

	//Trick: Use a temp variable and a tempStack that's descending(biggest E at top) 
	static void sortStackAscend(Stack<Integer> stack){
		if(stack.size() == 0) return; 

		Stack<Integer> tempStack = new Stack<Integer>();
		int temp;

		while(stack.size() != 0){
			if(tempStack.size() == 0){
				tempStack.push(stack.pop());
			} else if(stack.peek() > tempStack.peek()){
				tempStack.push(stack.pop());
			} else {
				temp = stack.pop(); //pop it off
				while(tempStack.size() != 0 && tempStack.peek() > temp){
					stack.push(tempStack.pop());
				}
				tempStack.push(temp);
			}
		}

		while(tempStack.size() != 0){
			stack.push(tempStack.pop());
		}
	}

	public static void main(String[] args){
		Stack<Integer> myStack = new Stack<Integer>();
		myStack.push(1);
		myStack.push(5);
		myStack.push(2);
		myStack.push(3);
		myStack.push(1);
		//myStack.print();
		sortStackAscend(myStack);
		while(myStack.size() != 0){
			System.out.println(myStack.pop());
		}

	}
}

//Trick: Use a timestamp to check which dog/cat to dequeue
class AnimalShelter2{
	LinkedList<Animal> dogs = new LinkedList<Animal>();
	LinkedList<Animal> cats = new LinkedList<Animal>();
	int time = 0;

	void enqueue(Animal a){
		time++;
		a.time = time;
		if(a.type.equals("dog")){
			dogs.add(a);
		} else if (a.type.equals("cat")){
			cats.add(a);
		} else {
			System.out.println("Not a dog or cat!");
		}
	}
	Animal dequeueAny(){
		if(dogs.peek().time < cats.peek().time){
			dequeueDog();
		} else{
			dequeueCat();
		}
		return null; //will never be reached 
	}
	Animal dequeueDog(){
		if(dogs.size() != 0){
			return dogs.poll(); 
		} 
		return null;
	}
	Animal dequeueCat(){
		if(cats.size() != 0){
			return cats.poll();
		}
		return null;
	}
}

//have to traverse through a linkedlist to remove a specific type of animal
class AnimalShelter{
	LinkedList<Animal> queue = new LinkedList<Animal>();

	void enqueue(Animal animal){
		queue.add(animal);
	}

	Animal dequeueAny(){
		return queue.remove();
	}

	//iterating through a linkedlist...
	Animal dequeueDog(){
		int i =0;
		for(Animal a: queue){
			i++;
			if(a.type.equals("dog")){
				return queue.remove(i);
			}
		}
		return null; //no dog in list
	}
	Animal dequeueCat(){
		for(int i =0; i< queue.size(); i++){
			if(queue.get(i).type.equals("cat")){
				return queue.remove(i);
			}
		}
		return null; //no cat in list
	}

}

class Animal{
	String type;
	int time;
	Animal(String type){
		this.type = type;
	}
}

//Trick: Invert first stack when we need to pop by putting it into the second stack
class QueueAsTwoStacks{
	Stack<Integer> stack = new Stack<Integer>();
	Stack<Integer> queue = new Stack<Integer>();

	void add(int data){
		stack.push(data);
	}

	int remove(){
		if(queue.size() == 0){
			while(stack.size() != 0){
				queue.push(stack.pop());
			}
		}
		return queue.pop();
	}
}

//Trick: Use a stack of stacks. If you need popAt(), use an arraylist of linkedlists
class StackPlates{
	Stack<Stack<Integer>> allPlates = new Stack<Stack<Integer>>();
	int maxHeight;

	StackPlates(int maxHeight){
		this.maxHeight = maxHeight;
	}

	void push(int calories){
		if(allPlates.isEmpty() || allPlates.peek().size() == maxHeight){
			Stack<Integer> newStack = new Stack<Integer>();
			newStack.push(calories);
			allPlates.push(newStack);
		}
		else { //if(allPlates.peek().size() < stackHeight)
			allPlates.peek().push(calories);
		} 
	}

	void print(){ //modifies original stack... be careful
		while(allPlates.size() != 0){
			while(allPlates.peek().size() != 0){
				System.out.println(allPlates.peek().pop());
			}
			allPlates.pop();
		}
	}

	int pop(){
		if(allPlates.isEmpty()){
			throw new EmptyStackException();
		} else if(allPlates.peek().size() > 1){
			//System.out.println("--");
			return allPlates.peek().pop();
		} else {
			int element = allPlates.peek().peek();
			allPlates.pop();
			return element;
		}
	}

	int popAt(int stackNum){
		//popping at a specific stack, we need a list of lists
		//stack of stacks do not seem to work
		//pretending we have an ArrayList of LinkLists...
		//LinkLists are really good because we can always represent a stack or queue with it
		//even the methods are very similar
		//also works with ArrayList of Deques! 
		//this begs the question, LinkList vs Queues to represent stacks/queues?
		// int element = allPlates.get(stackNum).pop();
		// for(int i = stackNum; i<allPlates.size()-1; i++){
		// 	//push onto previous linklist the last element of the next linklist
		// 	allPlates.get(i).push(allPlates.get(i+1).removeLast());
		// }
		// return element;
		return 0;
	}

}

//Trick: Use a second stack to store the past history of all the min values
//Min variable never changes unless we just popped it off our stack
class StackHasMinMethod{
	Stack<Integer> stack = new Stack<Integer>();
	int min = Integer.MAX_VALUE;
	Stack<Integer> minStack = new Stack<Integer>();

	void push(int val){
		if(val<min){
			min = val;
			minStack.push(min);
		}
		stack.push(val);
	}

	int pop(){
		if(!stack.empty()){
			if(stack.peek() == min){
				minStack.pop();
				min = minStack.peek();
			}
			return stack.pop();
		} else {
			throw new EmptyStackException();
		}
	}
	
	int min(){
		if(!stack.empty()){
			return min;
		} else{
			throw new EmptyStackException();
		}
	}
}

//[0, n/3), [n/3, 2n/3), [2n/3, n)
//Trick: Keep track of sizes of each stack with another array 
class ArrayOfNStacks{
	int numStacks = 3; //n
	int stackCapacity;
	int[] values; //store value of all stacks
	int[] sizes = new int[numStacks]; //store size of each individual stack

	ArrayOfNStacks(int stackCapacity){
		values = new int[stackCapacity * numStacks];
		this.stackCapacity = stackCapacity;
	}

	void push(int stackNum, int value){
		if(isFull(stackNum)){
			System.out.println("Sorry, you have reached stack overflow!");
		}
		sizes[stackNum]++;
		values[indexOfTop(stackNum)] = value;
	}

	int peek(int stackNum){
		if(isEmpty(stackNum)){
			throw new EmptyStackException();
		}
		return values[indexOfTop(stackNum)];
	}

	int pop(int stackNum, int value){
		if(!isEmpty(stackNum)){
			int temp = values[indexOfTop(stackNum)];
			values[indexOfTop(stackNum)] = 0;
			sizes[stackNum]--;
			return temp;
		} else {
			throw new EmptyStackException();
		}
	}

	boolean isEmpty(int stackNum){
		return sizes[stackNum] == 0;
	}

	boolean isFull(int stackNum){
		return sizes[stackNum] == stackCapacity;
	}

	int indexOfTop(int stackNum){
		int offset = stackCapacity*stackNum;
		int size = sizes[stackNum];
		return offset + size -1;
	}
}

File: Stacks and Queues/stackqueue.py
import sys
from collections import deque


class ThreeInOne:  # idea to use a wrapper class and keep a separate array for the sizes
    def __init__(self, stack_capacity):
        self.array = [None] * 3 * stack_capacity
        self.stack_capacity = stack_capacity
        self.sizes = [0] * 3

    def push(self, stack_num, value):
        if self.sizes[stack_num - 1] >= self.stack_capacity:
            raise Exception("Error, max stack stack_capacity!")
        self.sizes[stack_num - 1] += 1
        self.array[self.stack_capacity * (stack_num - 1) + self.sizes[stack_num - 1] - 1] = value

    def pop(self, stack_num):
        if self.sizes[stack_num - 1] == 0:
            raise Exception("Error, popping from empty stack!")
        value = self.array[self.stack_capacity * (stack_num - 1) + self.sizes[stack_num - 1] - 1]
        self.array[self.stack_capacity * (stack_num - 1) + self.sizes[stack_num - 1] - 1] = None
        self.sizes[stack_num - 1] -= 1
        return value


class MinStack:  # idea that you only need the history of the smallest values
    def __init__(self):
        self.stack = []
        self.min = sys.maxint
        self.min_array = []

    def push(self, value):
        self.stack.append(value)
        if value <= self.min:
            self.min = value
            self.min_array.append(value)

    def _min(self):
        if len(self.stack) == 0:
            raise Exception("Stack is empty!")
        return self.min_array[-1]

    def pop(self):
        if len(self.stack) == 0:
            raise Exception("Stack is empty!")
        value = self.stack.pop()
        if value == self.min_array[-1]:
            self.min_array.pop()
        return value


class SetOfStacks:  # trick to use array of deques for pop_at()
    def __init__(self, capacity):
        self.capacity = capacity
        self.arr_stack = [deque()]

    def push(self, value):
        if len(self.arr_stack[-1]) == self.capacity:
            self.arr_stack.append(deque([value]))  # cool deque literal
        else:
            self.arr_stack[-1].append(value)

    def pop(self):
        if len(self.arr_stack) == 0:
            raise Exception("Error, empty stack!")
        value = self.arr_stack[-1].pop()
        if len(self.arr_stack[-1]) == 0:
            self.arr_stack.pop()
        return value

    def pop_at(self, index):
        if index > len(self.arr_stack) - 1:
            raise Exception("Error, invalid index!")
        value = self.arr_stack[index].pop()
        for i in range(index + 1, len(self.arr_stack) - 1, 1):
            self.arr_stack[i - 1].append(self.arr_stack[i].popleft())
        return value


class QueueStack:
    def __init__(self):
        self.push_stack = []
        self.pop_stack = []

    def push(self, val):
        self.push_stack.append(val)

    def pop(self):
        if len(self.pop_stack) > 0:
            return self.pop_stack.pop()
        elif len(self.push_stack) == 0:
            raise Exception("Empty queue!")
        else:  # pop_stack is empty but push_stack still has values
            self.pop_stack = self.push_stack[::-1]  # reverse stack
            self.push_stack = []
            return self.pop_stack.pop()


def sort_stack(stack):  # idea to use a temp variable and temp stack then reverse
    if len(stack) == 0:
        raise Exception("Empty stack!")
    temp_stack, temp_val = [], 0
    temp_stack.append(stack.pop())
    while stack:
        if stack[-1] > temp_stack[-1]:
            temp_stack.append(stack.pop())
        else:
            temp_val = stack.pop()
            while temp_stack and temp_stack[-1] > temp_val:
                stack.append(temp_stack.pop())
            temp_stack.append(temp_val)

    while temp_stack:
        stack.append(temp_stack.pop())


class Dog:
    species = "Dog"

    def __init__(self, name):
        self.name = name


class Cat:
    species = "Cat"

    def __init__(self, name):
        self.name = name


class AnimalShelter:
    time_stamp = 0  # idea to use a timestamp and 2 separate queues

    def __init__(self):
        self.dogs = deque()
        self.cats = deque()

    def enque(self, animal):
        AnimalShelter.time_stamp += 1  # use of static variable
        if animal.species == "Dog":
            self.dogs.append((animal, AnimalShelter.time_stamp))  # deque of tuples
        elif animal.species == "Cat":
            self.cats.append((animal, AnimalShelter.time_stamp))
        else:
            raise Exception("Invalid animal!")

    def deque_dog(self):
        return self.dogs.popleft()[0]

    def deque_cat(self):
        return self.cats.popLeft()[0]

    def deque_any(self):
        if len(self.cats) == 0 and len(self.dogs) == 0:
            raise Exception("No animals!")
        elif len(self.cats) == 0:
            return self.dogs.popleft()[0]
        elif len(self.dogs) == 0:
            return self.cats.popleft()[0]
        elif self.cats[0][1] < self.dogs[0][1]:
            return self.cats.popleft()[0]
        else:
            return self.dogs.popleft()[0]

# d1 = Dog("A")
# d2 = Dog("B")
# c1 = Dog("C")
# c2 = Dog("D")
#
# shelter = AnimalShelter()
# shelter.enque(d1)
# shelter.enque(c1)
# shelter.enque(d2)
# shelter.enque(c2)
#
# print(shelter.dequeAny().name, shelter.dequeAny().name, shelter.dequeAny().name, shelter.dequeAny().name)
